### DISCUSSION TITLE
Not routing on first page load

### DISCUSSION THREAD

**Discussion by UbiquitousBear** [vercel/next.js#64568] [Help]
### Summary

When I load my app with a URI e.g. `/component/1`, I'd expect `component/[id]/index.tsx` to be rendered, using the page router (as I've set it up). However, on the first page load, the `index.tsx` for the app is rendered and not `component/[id]/index.tsx`. Subsequent `router.pushes` work as expected.

I have an app with a custom `_app.tsx`, which I'm using to wrap providers for different things. Is there something I'm missing to get the router to correctly render the right page? 

```
export default function App({ Component, pageProps }: AppProps) {
    return (
        <ClientSideRendering>
            <ThemedApp>
                <FlagsProvider>
                    <ApolloProvider client={graphqlclient}>
                        <div style={{backgroundColor: token('elevation.surface')}}>
                            <Navigation/>
                            <Component {...pageProps} />
                        </div>
                    </ApolloProvider>
                </FlagsProvider>
            </ThemedApp>
        </ClientSideRendering>
    )
}
```

**Comment by icyJoseph** [comment]
This could be many things...

Maybe you have a permanent redirect setup at some point? Then you need to clear your browser Cache and reload, try this https://help.codehs.com/en/articles/4951972-how-to-clear-your-browser-cache-and-hard-refresh


Otherwise, maybe we need to debug the `ClientSideRendering` component, and your file structure.

**Reply by UbiquitousBear** [reply]
I suspect it’s the `ClientSideRendering`, here is the code for it, to force the rendering client side.

```
import dynamic from 'next/dynamic';

type ClientOnlyProps = { children: JSX.Element };
const ClientOnly = (props: ClientOnlyProps) => {
    const { children } = props;
    return children;
};

export default dynamic(() => Promise.resolve(ClientOnly), {
    ssr: false,
});
```

The routes are not being cached in the browser. There are instances where routing doesn’t change the component but the URL in the address bar updates.

**Reply by icyJoseph** [reply]
Aha, I've never seen that pattern, where did you find it? 

That being said, try:

```tsx
export const ClientOnly = (props: ClientOnlyProps) => {
  const [isServer,setServer] = useState(true);
  useEffect(setServer,[]);

  if (isServer) return null;
  return props.children;
}
```

**Reply by UbiquitousBear** [reply]
@icyJoseph the original code came from [here](https://stackoverflow.com/questions/75842286/how-to-render-components-in-nextjs-13-only-on-client).

I've added 

```
'use client'
import {useEffect, useState} from "react";

type ClientOnlyProps = { children: JSX.Element };

const ClientSideRendering = (props: ClientOnlyProps) => {
    const [isServer,setServer] = useState(true);
    useEffect(() =>  setServer(false),[]);

    if (isServer) return null;
    return props.children;
}

export default ClientSideRendering
```

However, still not seeing the expected behaviour. If I remove the `<ClientSideRendering>` from `_app.tsx`, routing still isn't behaving, so maybe it's not that?

I've added a `console.log(Component)` to the `_app.tsx` and can confirm that the incorrect Component is being received: ie, the component defined in my app's `Index.tsx`.

**Reply by icyJoseph** [reply]
Yeah, then we should look at the file structure, or try to create a reproduction repository

**Reply by UbiquitousBear** [reply]
@icyJoseph before I push up some code, does this help? 

<img width="348" alt="image" src="https://github.com/vercel/next.js/assets/5007497/8f03b7a5-7769-4710-a297-7bdf9e16f428">

**Reply by GabenGar** [reply]
Why having this hack in the first place? If you 100% want to render something client-only then write a hook like this:
```typescript
export function useClient(): boolean {
   const [isClient, switchClient] = useState(false);
   
   useEffect(() => isClient(true), [])
   
   return isClient
}
```

And then check its return value within related components. Not sure why do you need pass props into it, since client/server environment doesn't depend on any application code.

**Reply by UbiquitousBear** [reply]
@GabenGar the whole app should be client rendered only; there was a package I was using which was misbehaving, however, that seems not to be the case anymore, so I think I'm safe to remove the workaround: one less component is easier.

**Reply by UbiquitousBear** [reply]
@icyJoseph Let me know if you want me to push up the code

**Reply by UbiquitousBear** [reply]
@icyJoseph  / @GabenGar any more ideas?
